### ISSUE TITLE
GradCAM images don't support backpropagation

### ISSUE THREAD

**Issue by DrumsnChocolate** [jacobgil/pytorch-grad-cam#266]
I am working on a project where we want to use GradCAM in a differentiable manner. That is, we are using a neural network to generate modified images which are meant to influence the Class Activation Maps of a classifier. I was surprised to see that, although the type annotation of `BaseCAM.get_cam_image`  suggests that the argument  `grads` has a `torch.Tensor` type, it in reality is a numpy array.  This prevents calculating gradients of the Class Activation Maps. (That is, I would like to compute how the inputs influence the values of the Class Activation Maps)
See the type definition below.

```
def get_cam_image(self,
                      input_tensor: torch.Tensor,
                      target_layer: torch.nn.Module,
                      targets: List[torch.nn.Module],
                      activations: torch.Tensor,
                      grads: torch.Tensor,
                      eigen_smooth: bool = False) -> np.ndarray:
```

Is this intentional?

**Comment by jacobgil** [comment]
Hi,
Not intentional at all, I just didn't think of this use case.

Since Grad-CAM uses gradients,
to compute gradients of CAM there would be second order gradients going on ?

I think it's safe to the functions to tensors.

**Comment by DrumsnChocolate** [comment]
I'm rewriting the methods where necessary for my project, and If I am able to produce a working implementation I'd be happy to contribute, but I am running into a slight issue.
The method `get_2d_projection` which is used in `BaseCAM.get_cam_image` makes use of a pytorch implementation of SVD: https://pytorch.org/docs/stable/generated/torch.linalg.svd.html
This decomposition seems fairly hard to compute gradients for, as indicated in the warning texts. Do you know if there are subclasses of `BaseCAM` that do not make use of SVD? That would really help me out.
